### ISSUE TITLE
False positive 'This comparison appears to be unintentional' narrowing string unions

### ISSUE THREAD

**Issue by dunhamsteve** [microsoft/TypeScript#55420]
### 🔎 Search Terms

ts2367 "have no overlap" string union

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about Common "Bugs" That Aren't Bugs
- Tested on current playground versions 3.3.3 through v5.3.0-dev.20230817

### ⏯ Playground Link

https://www.typescriptlang.org/play?ts=3.7.5#code/MYGwhgzhAEBiD29oG8BQ1oQC5iwUwC5oByeAOxAEsy9joAfE+AM2aproF4mLrb1ozRAAoAlCgEZKzaMKwALShAB02XHmgBCTt1K8O4gE54sAV0NlJ0BUuUAjMIbFXpsmyrX5oO3SzZ9icWhjMwsrYHIIeBA8ZRB4AHNhYmxKEBBocnZaUQEAXwEHJ3E0DAx3VRwvX1Zs4nzUAqA

### 💻 Code

```ts
class Foo {
  state: 'online' | 'offline' = 'online'
  foo() {
    if (this.state !== 'online') return
    this.bar()
    if (this.state === 'offline')  return
    console.log('still online')
  }
  bar() {
    this.state = 'offline'
  }
}
```

### 🙁 Actual behavior

Typescript reports a TS2367 error:

```
bug.ts:6:9 - error TS2367: This comparison appears to be unintentional because the types '"online"' and '"offline"' have no overlap.

6     if (this.state === 'offline')  return
          ~~~~~~~~~~~~~~~~~~~~~~~~
```

### 🙂 Expected behavior

Typescript should not report an error - the narrowing of `this.state` should be invalidated after the function call on line 5.  

### Additional information about the issue

Typescript behaves correctly if `this.state` has type `boolean`, but if the type is a union of strings, we get the error.

**Comment by MartinJohns** [comment]
Another duplicate of #9998.

**Comment by dunhamsteve** [comment]
Ahh, out of habit, I had put `!this.state` instead of `this.state === false` when I tried `boolean`, but it looks like error doesn't happen in that case, even without the function call. [Playground link](https://www.typescriptlang.org/play?ts=3.7.5#code/MYGwhgzhAEBiD29oG8BQ1oQC5iwUwC5oAjREPMAO2gF5osAnAVz3WgDNEAKAShTYwBLdtC5YAFoIgA6bLjzQAhDTqMWfBnixMGlAfUkziYBr33DRiiVNk58faJu279weJQjxy0kPADmXADk2IIgINDuIIKUeIE8bAC+bMamfGgYGNYycvi0HGAgEKwYSUlAA) 

I worked around it, so it isn't blocking me. And reading #9998 it sounds like it might not be desirable to fix the underlying issue.  Thanks for taking a look.
